<html><head></head><body><section data-pdf-bookmark="Chapter 18. FASTX Sampler: Randomly Subsampling Sequence Files" data-type="chapter" epub:type="chapter"><div class="chapter" id="ch18">&#13;
<h1><span class="label">Chapter 18. </span>FASTX Sampler: Randomly Subsampling Sequence Files</h1>&#13;
&#13;
&#13;
<p>Sequence datasets in genomics and metagenomics can get dauntingly large, requiring copious time and compute resources to analyze.<a data-primary="FASTX grep created" data-secondary="sampler" data-tertiary="about subsampling sequence files" data-type="indexterm" id="idm45963627124088"/><a data-primary="sequence file subsampling" data-secondary="about" data-type="indexterm" id="idm45963627122856"/><a data-primary="memory issues" data-secondary="sequence file subsampling" data-seealso="sequence file subsampling" data-type="indexterm" id="idm45963627121896"/><a data-primary="input for program" data-secondary="sequence file subsampling" data-seealso="sequence file subsampling" data-type="indexterm" id="idm45963627120648"/>&#13;
Many sequencers can produce tens of millions of reads per sample, and many experiments involve tens to hundreds of samples, each with multiple technical replicates resulting in gigabytes to terabytes of data.&#13;
Reducing the size of the input files by randomly subsampling sequences allows you to explore data more quickly.&#13;
In this chapter, I will show how to use Python’s &#13;
<span class="keep-together"><code>random</code></span> module to select some portion of the reads from FASTA/FASTQ sequence files.</p>&#13;
&#13;
<p>You will learn about:</p>&#13;
&#13;
<ul>&#13;
<li>&#13;
<p>Nondeterministic sampling</p>&#13;
</li>&#13;
</ul>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Getting Started" data-type="sect1"><div class="sect1" id="idm45963627116008">&#13;
<h1>Getting Started</h1>&#13;
&#13;
<p>The code and tests for this exercise are in the <em>18_fastx_sampler</em> directory.<a data-primary="FASTX grep created" data-secondary="sampler" data-tertiary="getting started" data-type="indexterm" id="idm45963627114232"/><a data-primary="sequence file subsampling" data-secondary="getting started" data-type="indexterm" id="idm45963627113016"/>&#13;
Start by copying the solution for a program called <code>sampler.py</code>:</p>&#13;
&#13;
<pre data-type="programlisting">$ cd 18_fastx_sampler/&#13;
$ cp solution.py sampler.py</pre>&#13;
&#13;
<p>The FASTA input files for testing this program will be generated by the <code>synth.py</code> program you wrote in <a data-type="xref" href="ch17.html#ch17">Chapter 17</a>.<a data-primary="synthesizing DNA via Markov chains" data-secondary="sequence file subsampling source files" data-type="indexterm" id="idm45963627108872"/><a data-primary="FASTX grep created" data-secondary="sampler" data-seealso="synthesizing DNA via Markov chains" data-tertiary="FASTA input files synthesized" data-type="indexterm" id="idm45963627107896"/><a data-primary="sequence file subsampling" data-secondary="FASTA input files synthesized" data-seealso="synthesizing DNA" data-type="indexterm" id="idm45963627106376"/>&#13;
If you didn’t finish writing that program, be sure to copy the solution to that filename before executing <strong><code>make fasta</code></strong> to create three FASTA files with 1K, 10K, and 100K reads, each between 75 and 200 bp in length, with filenames of <em>n1k.fa</em>, <em>n10k.fa</em>, and <em>n100k.fa</em>, respectively.&#13;
Use <code>seqmagique.py</code> to verify that the files are correct:</p>&#13;
&#13;
<pre data-type="programlisting">$ ../15_seqmagique/seqmagique.py tests/inputs/n1*&#13;
name                     min_len    max_len    avg_len    num_seqs&#13;
tests/inputs/n100k.fa         75        200     136.08      100000&#13;
tests/inputs/n10k.fa          75        200     136.13       10000&#13;
tests/inputs/n1k.fa           75        200     135.16        1000</pre>&#13;
&#13;
<p>Run <code>sampler.py</code> to select the default of 10% of the sequences from the smallest file.&#13;
If you use a random seed of <code>1</code>, you should get 95 reads:</p>&#13;
&#13;
<pre data-type="programlisting">$ ./sampler.py -s 1 tests/inputs/n1k.fa&#13;
Wrote 95 sequences from 1 file to directory "out"</pre>&#13;
&#13;
<p>The results can be found in a file called <em>n1k.fa</em> in an output directory named <em>out</em>.&#13;
One way to verify this is to use <strong><code>grep -c</code></strong> to count how many times it finds the symbol <code>&gt;</code> at the start of each record:</p>&#13;
&#13;
<pre data-type="programlisting">$ grep -c '&gt;' out/n1k.fa&#13;
95</pre>&#13;
&#13;
<p>Note that there is a pernicious error waiting for you if you happen to forget the quotes around <code>&gt;</code>:</p>&#13;
&#13;
<pre data-type="programlisting">$ grep -c &gt; out/n1k.fa&#13;
usage: grep [-abcDEFGHhIiJLlmnOoqRSsUVvwxZ] [-A num] [-B num] [-C[num]]&#13;
	[-e pattern] [-f file] [--binary-files=value] [--color=when]&#13;
	[--context[=num]] [--directories=action] [--label] [--line-buffered]&#13;
	[--null] [pattern] [file ...]</pre>&#13;
&#13;
<p>Wait, what just happened?&#13;
Remember that <code>&gt;</code> is the <code>bash</code> operator to <em>redirect</em> the <code>STDOUT</code> from one program into a file.<a data-primary="file output via redirect output (&gt;)" data-type="indexterm" id="idm45963627091688"/><a data-primary="output from program" data-secondary="&gt; redirect output" data-type="indexterm" id="idm45963627091000"/><a data-primary="&gt; (redirect output)" data-primary-sortas="# redirect output" data-type="indexterm" id="idm45963627090056"/>&#13;
In the preceding command, I ran <code>grep</code> without enough arguments and redirected the output into <em>out/n1k.fa</em>.&#13;
The output you see is the usage that is printed to <code>STDERR</code>.&#13;
Nothing was printed to <code>STDOUT</code>, so this null output overwrote the <em>out/n1k.fa</em> file and it is now empty:</p>&#13;
&#13;
<pre data-type="programlisting">$ wc out/n1k.fa&#13;
       0       0       0 out/n1k.fa</pre>&#13;
&#13;
<p>I point this out because I have lost several sequence files due to this gem.&#13;
The data has been lost permanently, so I must rerun the earlier command to regenerate the file.&#13;
After doing that, I recommend instead that you use <code>seqmagique.py</code> to verify the <span class="keep-together">contents:</span></p>&#13;
&#13;
<pre data-type="programlisting">$ ../15_seqmagique/seqmagique.py out/n1k.fa&#13;
name          min_len    max_len    avg_len    num_seqs&#13;
out/n1k.fa         75        200     128.42          95</pre>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Reviewing the Program Parameters" data-type="sect2"><div class="sect2" id="idm45963627083128">&#13;
<h2>Reviewing the Program Parameters</h2>&#13;
&#13;
<p>This is a fairly complex program with many options.<a data-primary="FASTX grep created" data-secondary="sampler" data-tertiary="program parameters" data-type="indexterm" id="idm45963627081784"/><a data-primary="sequence file subsampling" data-secondary="program parameters" data-type="indexterm" id="idm45963627080568"/><a data-primary="parameters" data-secondary="sequence file subsampling" data-type="indexterm" id="idm45963627079608"/><a data-primary="arguments" data-secondary="sequence file subsampling" data-type="indexterm" id="idm45963627078648"/>&#13;
Run the <code>sampler.py</code> program to request help.&#13;
Notice that the only required arguments are the input files, as all the options are set to reasonable defaults:</p>&#13;
&#13;
<pre data-type="programlisting">$ ./sampler.py -h&#13;
usage: sampler.py [-h] [-f format] [-p reads] [-m max] [-s seed] [-o DIR]&#13;
                  FILE [FILE ...]&#13;
&#13;
Probabilistically subset FASTA files&#13;
&#13;
positional arguments:&#13;
  FILE                  Input FASTA/Q file(s) <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO1-1" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO1-1"><img alt="1" src="assets/1.png"/></a>&#13;
&#13;
optional arguments:&#13;
  -h, --help            show this help message and exit&#13;
  -f format, --format format&#13;
                        Input file format (default: fasta) <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO1-2" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO1-2"><img alt="2" src="assets/2.png"/></a>&#13;
  -p reads, --percent reads&#13;
                        Percent of reads (default: 0.1) <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO1-3" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO1-3"><img alt="3" src="assets/3.png"/></a>&#13;
  -m max, --max max     Maximum number of reads (default: 0) <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO1-4" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO1-4"><img alt="4" src="assets/4.png"/></a>&#13;
  -s seed, --seed seed  Random seed value (default: None) <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO1-5" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO1-5"><img alt="5" src="assets/5.png"/></a>&#13;
  -o DIR, --outdir DIR  Output directory (default: out) <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO1-6" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO1-6"><img alt="6" src="assets/6.png"/></a></pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO1-1" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO1-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>One or more <em>FASTA</em> or <em>FASTQ</em> files are required.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO1-2" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO1-2"><img alt="2" src="assets/2.png"/></a></dt>&#13;
<dd><p>The default sequence format of the input files is <em>FASTA</em>.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO1-3" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO1-3"><img alt="3" src="assets/3.png"/></a></dt>&#13;
<dd><p>By default, the program will select 10% of the reads.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO1-4" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO1-4"><img alt="4" src="assets/4.png"/></a></dt>&#13;
<dd><p>This option will stop sampling when a given maximum is reached.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO1-5" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO1-5"><img alt="5" src="assets/5.png"/></a></dt>&#13;
<dd><p>This option sets the random seed to reproduce the selections.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO1-6" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO1-6"><img alt="6" src="assets/6.png"/></a></dt>&#13;
<dd><p>The default directory for the output files is <em>out</em>.</p></dd>&#13;
</dl>&#13;
&#13;
<p>As with previous programs, the program will reject invalid or unreadable input files, and the random seed argument must be an integer value.&#13;
The <code>-p|--percent</code> option should be a floating-point value between 0 and 1 (not inclusive), and the program will reject anything outside of that range.&#13;
I manually validate this argument and use <code>parser.error()</code>, as in Chapters <a data-type="xref" data-xrefstyle="select:labelnumber" href="ch04.html#ch04">4</a> and <a data-type="xref" data-xrefstyle="select:labelnumber" href="ch09.html#ch09">9</a>:</p>&#13;
&#13;
<pre data-type="programlisting">$ ./sampler.py -p 3 tests/inputs/n1k.fa&#13;
usage: sampler.py [-h] [-f format] [-p reads] [-m max] [-s seed] [-o DIR]&#13;
                  FILE [FILE ...]&#13;
sampler.py: error: --percent "3.0" must be between 0 and 1</pre>&#13;
&#13;
<p>The <code>-f|--format</code> option will only accept the values <code>fasta</code> or <code>fastq</code> and will default to the first.&#13;
I use the <code>choices</code> option with <code>argparse</code>, as in Chapters <a data-type="xref" data-xrefstyle="select:labelnumber" href="ch15.html#ch15">15</a> and <a data-type="xref" data-xrefstyle="select:labelnumber" href="ch16.html#ch16">16</a>, to automatically reject unwanted values.&#13;
For instance, the program will reject a value of <code>fastb</code>:</p>&#13;
&#13;
<pre data-type="programlisting">$ ./sampler.py -f fastb tests/inputs/n1k.fa&#13;
usage: sampler.py [-h] [-f format] [-p reads] [-m max] [-s seed] [-o DIR]&#13;
                  FILE [FILE ...]&#13;
sampler.py: error: argument -f/--format: invalid choice:&#13;
'fastb' (choose from 'fasta', 'fastq')</pre>&#13;
&#13;
<p>Finally, the <code>-m|--max</code> option defaults to <code>0</code>, meaning that the program will sample about <code>--percent</code> of the reads with no upper limit.&#13;
Practically speaking, you may have input files with tens of millions of reads but you only want at most 100K from each.&#13;
Use this option to halt sampling when the desired number is reached.&#13;
For instance, I can use <code>-m 30</code> to stop sampling at 30 reads:</p>&#13;
&#13;
<pre data-type="programlisting">$ ./sampler.py -m 30 -s 1 tests/inputs/n1k.fa&#13;
  1: n1k.fa&#13;
Wrote 30 sequences from 1 file to directory "out"</pre>&#13;
&#13;
<p>When you think you understand how the program should work, start over with your version:</p>&#13;
&#13;
<pre data-type="programlisting">$ new.py -fp 'Probabilistically subset FASTA files' sampler.py&#13;
Done, see new script "sampler.py".</pre>&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Defining the Parameters" data-type="sect2"><div class="sect2" id="idm45963627029048">&#13;
<h2>Defining the Parameters</h2>&#13;
&#13;
<p>The arguments to the program contain many different data types, which I represent with the following class:<a data-primary="FASTX grep created" data-secondary="sampler" data-tertiary="program parameters defined" data-type="indexterm" id="idm45963627027656"/><a data-primary="parameters" data-secondary="sequence file subsampling" data-tertiary="parameters defined" data-type="indexterm" id="idm45963627026424"/><a data-primary="arguments" data-secondary="sequence file subsampling" data-tertiary="parameters defined" data-type="indexterm" id="idm45963627025192"/></p>&#13;
&#13;
<pre data-type="programlisting">class Args(NamedTuple):&#13;
    """ Command-line arguments """&#13;
    files: List[TextIO] <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO2-1" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO2-1"><img alt="1" src="assets/1.png"/></a>&#13;
    file_format: str <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO2-2" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO2-2"><img alt="2" src="assets/2.png"/></a>&#13;
    percent: float <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO2-3" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO2-3"><img alt="3" src="assets/3.png"/></a>&#13;
    max_reads: int <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO2-4" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO2-4"><img alt="4" src="assets/4.png"/></a>&#13;
    seed: Optional[int] <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO2-5" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO2-5"><img alt="5" src="assets/5.png"/></a>&#13;
    outdir: str <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO2-6" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO2-6"><img alt="6" src="assets/6.png"/></a></pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO2-1" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO2-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>The files are a list of open filehandles.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO2-2" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO2-2"><img alt="2" src="assets/2.png"/></a></dt>&#13;
<dd><p>The input file format is a string.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO2-3" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO2-3"><img alt="3" src="assets/3.png"/></a></dt>&#13;
<dd><p>The percentage of reads is a floating-point value.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO2-4" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO2-4"><img alt="4" src="assets/4.png"/></a></dt>&#13;
<dd><p>The maximum number of reads is an integer.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO2-5" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO2-5"><img alt="5" src="assets/5.png"/></a></dt>&#13;
<dd><p>The random seed value is an optional integer.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO2-6" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO2-6"><img alt="6" src="assets/6.png"/></a></dt>&#13;
<dd><p>The output directory name is a string.</p></dd>&#13;
</dl>&#13;
&#13;
<p>Here is how I define the arguments with <code>argparse</code>:</p>&#13;
&#13;
<pre data-type="programlisting">def get_args() -&gt; Args:&#13;
    parser = argparse.ArgumentParser(&#13;
        description='Probabilistically subset FASTA files',&#13;
        formatter_class=argparse.ArgumentDefaultsHelpFormatter)&#13;
&#13;
    parser.add_argument('file',&#13;
                        metavar='FILE',&#13;
                        type=argparse.FileType('r'), <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-1" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO3-1"><img alt="1" src="assets/1.png"/></a>&#13;
                        nargs='+',&#13;
                        help='Input FASTA/Q file(s)')&#13;
&#13;
    parser.add_argument('-f',&#13;
                        '--format',&#13;
                        help='Input file format',&#13;
                        metavar='format',&#13;
                        type=str,&#13;
                        choices=['fasta', 'fastq'], <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-2" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO3-2"><img alt="2" src="assets/2.png"/></a>&#13;
                        default='fasta')&#13;
&#13;
    parser.add_argument('-p',&#13;
                        '--percent',&#13;
                        help='Percent of reads',&#13;
                        metavar='reads',&#13;
                        type=float, <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-3" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO3-3"><img alt="3" src="assets/3.png"/></a>&#13;
                        default=.1)&#13;
&#13;
    parser.add_argument('-m',&#13;
                        '--max',&#13;
                        help='Maximum number of reads',&#13;
                        metavar='max',&#13;
                        type=int,&#13;
                        default=0) <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-4" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO3-4"><img alt="4" src="assets/4.png"/></a>&#13;
&#13;
    parser.add_argument('-s',&#13;
                        '--seed',&#13;
                        help='Random seed value',&#13;
                        metavar='seed',&#13;
                        type=int,&#13;
                        default=None) <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-5" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO3-5"><img alt="5" src="assets/5.png"/></a>&#13;
&#13;
    parser.add_argument('-o',&#13;
                        '--outdir',&#13;
                        help='Output directory',&#13;
                        metavar='DIR',&#13;
                        type=str,&#13;
                        default='out') <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-6" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO3-6"><img alt="6" src="assets/6.png"/></a>&#13;
&#13;
    args = parser.parse_args()&#13;
&#13;
    if not 0 &lt; args.percent &lt; 1: <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-7" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO3-7"><img alt="7" src="assets/7.png"/></a>&#13;
        parser.error(f'--percent "{args.percent}" must be between 0 and 1')&#13;
&#13;
    if not os.path.isdir(args.outdir): <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-8" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO3-8"><img alt="8" src="assets/8.png"/></a>&#13;
        os.makedirs(args.outdir)&#13;
&#13;
    return Args(files=args.file, <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-9" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO3-9"><img alt="9" src="assets/9.png"/></a>&#13;
                file_format=args.format,&#13;
                percent=args.percent,&#13;
                max_reads=args.max,&#13;
                seed=args.seed,&#13;
                outdir=args.outdir)</pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO3-1" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>Define the file inputs as one or more readable text files.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO3-2" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-2"><img alt="2" src="assets/2.png"/></a></dt>&#13;
<dd><p>Use <code>choices</code> to restrict the file formats and default to <code>fasta</code>.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO3-3" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-3"><img alt="3" src="assets/3.png"/></a></dt>&#13;
<dd><p>The percent argument is a floating-point value with a default of 10%.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO3-4" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-4"><img alt="4" src="assets/4.png"/></a></dt>&#13;
<dd><p>The maximum number of reads should be an integer with a default of <code>0</code>.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO3-5" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-5"><img alt="5" src="assets/5.png"/></a></dt>&#13;
<dd><p>The random seed value is optional but should be a valid integer if present.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO3-6" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-6"><img alt="6" src="assets/6.png"/></a></dt>&#13;
<dd><p>The output directory is a string with a default value of <code>out</code>.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO3-7" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-7"><img alt="7" src="assets/7.png"/></a></dt>&#13;
<dd><p>Verify that the percentage is between 0 and 1.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO3-8" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-8"><img alt="8" src="assets/8.png"/></a></dt>&#13;
<dd><p>Create the output directory if it does not exist.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO3-9" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO3-9"><img alt="9" src="assets/9.png"/></a></dt>&#13;
<dd><p>Return the <code>Args</code> object.</p></dd>&#13;
</dl>&#13;
&#13;
<p>Note that, while the program will accept both FASTA and FASTQ inputs, it should only write FASTA-formatted output files.</p>&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Nondeterministic Sampling" data-type="sect2"><div class="sect2" id="idm45963626946568">&#13;
<h2>Nondeterministic Sampling</h2>&#13;
&#13;
<p>Since there is no inherent ordering to sequences, one might be tempted to take the first however many sequences indicated by the user.<a data-primary="FASTX grep created" data-secondary="sampler" data-tertiary="nondeterministic sampling" data-type="indexterm" id="idm45963626945144"/><a data-primary="sequence file subsampling" data-secondary="nondeterministic sampling" data-type="indexterm" id="idm45963626943912"/><a data-primary="nondeterministic sampling" data-type="indexterm" id="idm45963626942936"/><a data-primary="sampling, nondeterministic" data-type="indexterm" id="idm45963626942248"/>&#13;
For instance, I could use <code>head</code> to select some number of lines from each file.&#13;
This would work on FASTQ files as long as this number is a divisor of four, but such an approach could create an invalid file for most any other sequence format, like multiline FASTA or Swiss-Prot.</p>&#13;
&#13;
<p>I’ve shown several programs that read and select sequences from files, so I could repurpose one of those to select records until the desired number is reached.&#13;
When I was first asked to write this program by my boss, I did exactly this.&#13;
The output was unusable, however, because I had not realized that the input was a synthetic dataset created by a colleague to simulate a <em>metagenome</em>, which is an environmental sample comprised of unknown organisms.<a data-primary="metagenome simulated" data-type="indexterm" id="idm45963626939400"/><a data-primary="synthesizing DNA via Markov chains" data-secondary="metagenome simulated" data-type="indexterm" id="idm45963626938728"/><a data-primary="input for program" data-secondary="metagenome simulated" data-type="indexterm" id="idm45963626937768"/>&#13;
The input file was created by concatenating various reads from known genomes so that, for instance, the first 10K reads were from a bacteria, the next 10K from another bacteria, the next 10K from representative archaea, the next 10K from viruses, and so on.&#13;
Taking just the first <em>N</em> records failed to include the diversity of the input.&#13;
It was not only a boring program to write, but worse, it always generated the same output and so could not be used to generate different subsamples.&#13;
This is an example of a <em>deterministic</em> program because the outputs were always the same given the same inputs.<a data-primary="deterministic program" data-type="indexterm" id="idm45963626935224"/></p>&#13;
&#13;
<p>Since I needed to find a way to randomly select some percentage of the reads, my first thought was to count the reads so I could figure out how many would be, for instance, 10%.&#13;
To do this, I stored all the sequences in a list, used <code>len()</code> to figure out how many were present, and then randomly selected 10% of the numbers in this range.&#13;
While such an approach might work for very small input files, I hope you can see how this fails to scale in any meaningful way.<a data-primary="memory issues" data-secondary="input file analysis" data-type="indexterm" id="idm45963626933224"/><a data-primary="scalability" data-secondary="input file analysis" data-seealso="memory issues" data-type="indexterm" id="idm45963626932280"/>&#13;
It’s not uncommon to encounter input files with tens of millions of reads.&#13;
Keeping all that data in something like a Python list could easily require more memory than is available on a machine.</p>&#13;
&#13;
<p>Eventually, I settled on a solution that reads one sequence at a time and randomly decides whether to choose or reject it.&#13;
That is, for each sequence I randomly select a number from a <em>continuous uniform distribution</em> between 0 and 1, meaning that all values in this range are equally likely to be selected.<a data-primary="continuous uniform distribution" data-type="indexterm" id="idm45963626929560"/><a data-primary="nondeterministic sampling" data-secondary="continuous uniform distribution" data-type="indexterm" id="idm45963626928872"/>&#13;
If that number is less than or equal to the given percentage, I select the read.&#13;
This approach only ever holds one sequence record in memory at a time, and so should scale at least linearly or <em>O(n)</em>.</p>&#13;
&#13;
<p>To demonstrate the selection process, I’ll import the <code>random</code> module and select a number between 0 and 1 using the <code>random.random()</code> function:<a data-primary="random() function" data-secondary="random()" data-type="indexterm" id="idm45963626925688"/></p>&#13;
&#13;
<pre data-type="programlisting">&gt;&gt;&gt; import random&#13;
&gt;&gt;&gt; random.random()&#13;
0.465289867914331</pre>&#13;
&#13;
<p>It’s unlikely that you’ll get the same number as I do.&#13;
We have to agree on a seed to both produce the same value.<a data-primary="random() function" data-secondary="seed()" data-type="indexterm" id="idm45963626923448"/>&#13;
Use the integer <code>1</code>, and you should get this number:</p>&#13;
&#13;
<pre data-type="programlisting">&gt;&gt;&gt; random.seed(1)&#13;
&gt;&gt;&gt; random.random()&#13;
0.13436424411240122</pre>&#13;
<div data-type="note" epub:type="note">&#13;
<p>The <code>random.random()</code> function uses a uniform distribution. The <code>random</code> module can also sample from other distributions, like normal or Gaussian. Consult <code>help(random)</code> to see these other functions and how to use them.<a data-primary="nondeterministic sampling" data-secondary="random.random() uniform distribution" data-type="indexterm" id="idm45963626918632"/></p>&#13;
</div>&#13;
&#13;
<p>As I iterate through the sequences in each file, I use this function to select a number.&#13;
If the number is less than or equal to the selected percentage, I want to write the sequence to an output file.&#13;
That is, the <code>random.random()</code> function ought to produce a number less than or equal to <code>.10</code> about 10% of the time.&#13;
In this way, I’m using a <em>nondeterministic</em> approach to sampling because the selected reads will be different each time I run the program (assuming I do not set a random seed).&#13;
This allows me to generate many different subsamples from the same input file, which could prove useful in generating technical replicates for analyses.</p>&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Structuring the Program" data-type="sect2"><div class="sect2" id="idm45963626914840">&#13;
<h2>Structuring the Program</h2>&#13;
&#13;
<p>You may feel somewhat overwhelmed by the complexity of this program, so I’ll present pseudocode you may find helpful:<a data-primary="FASTX grep created" data-secondary="sampler" data-tertiary="structuring the program" data-type="indexterm" id="idm45963626913448"/><a data-primary="sequence file subsampling" data-secondary="structuring the program" data-type="indexterm" id="idm45963626912232"/></p>&#13;
&#13;
<pre data-type="programlisting">set the random seed&#13;
iterate through each input file&#13;
    set the output filename to output directory plus the input file's basename&#13;
    open the output filehandle&#13;
    initialize a counter for how many records have been taken&#13;
&#13;
    iterate through each record of the input file&#13;
        select a random number between 0 and 1&#13;
        if this number is less than or equal to the percent&#13;
            write the sequence in FASTA format to the output filehandle&#13;
            increment the counter for taking records&#13;
&#13;
        if there is a max number of records and the number taken is equal&#13;
            leave the loop&#13;
&#13;
    close the output filehandle&#13;
&#13;
print how many sequences were taken from how many files and the output location</pre>&#13;
&#13;
<p>I would encourage you to run the <code>solution.py</code> program on one file and then on several files and try to reverse-engineer the output.&#13;
Keep running the test suite to ensure your program is on track.</p>&#13;
&#13;
<p>I hope you can see how similar in structure this program is to many previous programs that process some number of input files and create some output.&#13;
For example, in <a data-type="xref" href="ch02.html#ch02">Chapter 2</a> you processed files of DNA sequences to produce files of RNA sequences in an output directory.&#13;
In <a data-type="xref" href="ch15.html#ch15">Chapter 15</a>, you processed sequence files to produce a summary table of statistics.&#13;
In <a data-type="xref" href="ch16.html#ch16">Chapter 16</a>, you processed sequence files to select those records matching a pattern and wrote the selected sequences to an output file.&#13;
The programs in Chapters <a data-type="xref" data-xrefstyle="select:labelnumber" href="ch02.html#ch02">2</a> and <a data-type="xref" data-xrefstyle="select:labelnumber" href="ch16.html#ch16">16</a> most closely resemble what you need to do here, so I recommend you borrow from those solutions.</p>&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Solutions" data-type="sect1"><div class="sect1" id="idm45963626902360">&#13;
<h1>Solutions</h1>&#13;
&#13;
<p>I want to share two versions of a solution.&#13;
The first works to solve the exact problem as described.&#13;
The second solution goes beyond the original requirements because I want to show you how to overcome two common problems you may face in processing large bioinformatics datasets, namely opening too many filehandles and reading compressed files.</p>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Solution 1: Reading Regular Files" data-type="sect2"><div class="sect2" id="idm45963626900568">&#13;
<h2>Solution 1: Reading Regular Files</h2>&#13;
&#13;
<p>If you are dealing with a limited number of uncompressed input files, the following solution is appropriate:<a data-primary="FASTX grep created" data-secondary="sampler" data-tertiary="solution 1 regular file input" data-type="indexterm" id="idm45963626899160"/><a data-primary="sequence file subsampling" data-secondary="solution 1 regular file input" data-type="indexterm" id="idm45963626897928"/><a data-primary="file input" data-secondary="sequence file subsampling" data-type="indexterm" id="idm45963626896952"/></p>&#13;
&#13;
<pre data-type="programlisting">def main() -&gt; None:&#13;
    args = get_args()&#13;
    random.seed(args.seed) <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-1" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO4-1"><img alt="1" src="assets/1.png"/></a>&#13;
&#13;
    total_num = 0 <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-2" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO4-2"><img alt="2" src="assets/2.png"/></a>&#13;
    for i, fh in enumerate(args.files, start=1): <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-3" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO4-3"><img alt="3" src="assets/3.png"/></a>&#13;
        basename = os.path.basename(fh.name)&#13;
        out_file = os.path.join(args.outdir, basename)&#13;
        print(f'{i:3}: {basename}') <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-4" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO4-4"><img alt="4" src="assets/4.png"/></a>&#13;
&#13;
        out_fh = open(out_file, 'wt') <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-5" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO4-5"><img alt="5" src="assets/5.png"/></a>&#13;
        num_taken = 0 <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-6" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO4-6"><img alt="6" src="assets/6.png"/></a>&#13;
&#13;
        for rec in SeqIO.parse(fh, args.file_format): <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-7" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO4-7"><img alt="7" src="assets/7.png"/></a>&#13;
            if random.random() &lt;= args.percent: <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-8" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO4-8"><img alt="8" src="assets/8.png"/></a>&#13;
                num_taken += 1&#13;
                SeqIO.write(rec, out_fh, 'fasta')&#13;
&#13;
            if args.max_reads and num_taken == args.max_reads: <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-9" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO4-9"><img alt="9" src="assets/9.png"/></a>&#13;
                break&#13;
&#13;
        out_fh.close() <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-10" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO4-10"><img alt="10" src="assets/10.png"/></a>&#13;
        total_num += num_taken&#13;
&#13;
    num_files = len(args.files) <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-11" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO4-11"><img alt="11" src="assets/11.png"/></a>&#13;
    print(f'Wrote {total_num:,} sequence{"" if total_num == 1 else "s"} '&#13;
          f'from {num_files:,} file{"" if num_files == 1 else "s"} '&#13;
          f'to directory "{args.outdir}"')</pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO4-1" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>Set the random seed, if present. The default <code>None</code> value will be the same as not setting the seed.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO4-2" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-2"><img alt="2" src="assets/2.png"/></a></dt>&#13;
<dd><p>Initialize a variable to note the total number of sequences selected.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO4-3" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-3"><img alt="3" src="assets/3.png"/></a></dt>&#13;
<dd><p>Iterate through each input filehandle.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO4-4" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-4"><img alt="4" src="assets/4.png"/></a></dt>&#13;
<dd><p>Construct the output filename by joining the output directory with the file’s basename.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO4-5" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-5"><img alt="5" src="assets/5.png"/></a></dt>&#13;
<dd><p>Open the output filehandle for writing text.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO4-6" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-6"><img alt="6" src="assets/6.png"/></a></dt>&#13;
<dd><p>Initialize a variable to note how many sequences have been taken from this file.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO4-7" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-7"><img alt="7" src="assets/7.png"/></a></dt>&#13;
<dd><p>Iterate through each sequence record in the input file.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO4-8" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-8"><img alt="8" src="assets/8.png"/></a></dt>&#13;
<dd><p>If the record is randomly selected, increment the counter and write the sequence to the output file.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO4-9" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-9"><img alt="9" src="assets/9.png"/></a></dt>&#13;
<dd><p>If there is a maximum limit defined and the number of records selected is equal to this, exit the inner <code>for</code> loop.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO4-10" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-10"><img alt="10" src="assets/10.png"/></a></dt>&#13;
<dd><p>Close the output filehandle and increment the number of total records taken.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO4-11" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO4-11"><img alt="11" src="assets/11.png"/></a></dt>&#13;
<dd><p>Note the number of files processed and inform the user of the final status.</p></dd>&#13;
</dl>&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Solution 2: Reading a Large Number of Compressed Files" data-type="sect2"><div class="sect2" id="idm45963626842008">&#13;
<h2>Solution 2: Reading a Large Number of Compressed Files</h2>&#13;
&#13;
<p>The original problem does not involve reading compressed files, but you will often find that data will be stored in this way to save bandwidth in transferring data and disk space in storing it.<a data-primary="sequence file subsampling" data-secondary="solution 2 compressed file input" data-type="indexterm" id="idm45963626840504"/><a data-primary="FASTX grep created" data-secondary="sampler" data-tertiary="solution 2 compressed file input" data-type="indexterm" id="idm45963626839528"/><a data-primary="file input" data-secondary="sequence file subsampling" data-tertiary="compressed file input" data-type="indexterm" id="idm45963626838296"/><a data-primary="file input" data-secondary="compressed files" data-type="indexterm" id="idm45963626837064"/>&#13;
Python can directly read files compressed with tools like <code>zip</code> and <code>gzip</code>, so it’s not necessary to uncompress the input files before processing.</p>&#13;
&#13;
<p>Additionally, if you are processing hundreds to thousands of input files, you will find that using <code>type=argparse.FileType()</code> will cause your program to fail because you may exceed the maximum number of open files your operating system will allow.&#13;
In that case, you should declare <code>Args.files</code> as <code>List[str]</code> and create the parameter like this:</p>&#13;
&#13;
<pre data-type="programlisting">parser.add_argument('file',&#13;
                    metavar='FILE',&#13;
                    type=str, <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO5-1" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO5-1"><img alt="1" src="assets/1.png"/></a>&#13;
                    nargs='+',&#13;
                    help='Input FASTA/Q file(s)')</pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO5-1" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO5-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>Set the parameter type to one or more string values.</p></dd>&#13;
</dl>&#13;
&#13;
<p>This means you will have to validate the input files yourself, which you can do in the <code>get_args()</code> function, like so:</p>&#13;
&#13;
<pre data-type="programlisting">if bad_files := [file for file in args.file if not os.path.isfile(file)]: <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO6-1" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO6-1"><img alt="1" src="assets/1.png"/></a>&#13;
    parser.error(f'Invalid file: {", ".join(bad_files)}') <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO6-2" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO6-2"><img alt="2" src="assets/2.png"/></a></pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO6-1" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO6-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>Find all the arguments that are not valid files.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO6-2" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO6-2"><img alt="2" src="assets/2.png"/></a></dt>&#13;
<dd><p>Use <code>parser.error()</code> to report the bad inputs.</p></dd>&#13;
</dl>&#13;
&#13;
<p>The <code>main()</code> processing needs to change slightly as now <code>args.files</code> will be a list of strings.&#13;
You will need to open the filehandles yourself with <code>open()</code>, and this is the crucial change to your program needed to handle compressed files.&#13;
I will use a simple heuristic that examines the file extension for <code>.gz</code> to determine if a file is zipped and will instead use the <code>gzip.open()</code> function to open it:</p>&#13;
&#13;
<pre data-type="programlisting">def main() -&gt; None:&#13;
    args = get_args()&#13;
    random.seed(args.seed)&#13;
&#13;
    total_num = 0&#13;
    for i, file in enumerate(args.files, start=1): <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO7-1" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO7-1"><img alt="1" src="assets/1.png"/></a>&#13;
        basename = os.path.basename(file)&#13;
        out_file = os.path.join(args.outdir, basename)&#13;
        print(f'{i:3}: {basename}')&#13;
&#13;
        ext = os.path.splitext(basename)[1] <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO7-2" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO7-2"><img alt="2" src="assets/2.png"/></a>&#13;
        fh = gzip.open(file, 'rt') if ext == '.gz' else open(file, 'rt') <a class="co" href="#callout_fastx_sampler__randomly_subsampling_sequence_files_CO7-3" id="co_fastx_sampler__randomly_subsampling_sequence_files_CO7-3"><img alt="3" src="assets/3.png"/></a>&#13;
        out_fh = open(out_file, 'wt')&#13;
        num_taken = 0&#13;
&#13;
        for rec in SeqIO.parse(fh, args.file_format):&#13;
            if random.random() &lt;= args.percent:&#13;
                num_taken += 1&#13;
                SeqIO.write(rec, out_fh, 'fasta')&#13;
&#13;
            if args.max_reads and num_taken == args.max_reads:&#13;
                break&#13;
&#13;
        out_fh.close()&#13;
        total_num += num_taken&#13;
&#13;
    num_files = len(args.files)&#13;
    print(f'Wrote {total_num:,} sequence{"" if total_num == 1 else "s"} '&#13;
          f'from {num_files:,} file{"" if num_files == 1 else "s"} '&#13;
          f'to directory "{args.outdir}".')</pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO7-1" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO7-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p><code>args.files</code> is now a list of strings, not filehandles.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO7-2" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO7-2"><img alt="2" src="assets/2.png"/></a></dt>&#13;
<dd><p>Get the file extension.</p></dd>&#13;
<dt><a class="co" href="#co_fastx_sampler__randomly_subsampling_sequence_files_CO7-3" id="callout_fastx_sampler__randomly_subsampling_sequence_files_CO7-3"><img alt="3" src="assets/3.png"/></a></dt>&#13;
<dd><p>If the file extension is <code>.gz</code>, use <code>gzip.open()</code> to open the file; otherwise, use the normal <code>open()</code> function.</p></dd>&#13;
</dl>&#13;
&#13;
<p>Finally, there are times when <code>nargs='+'</code> will also not work.&#13;
For one project, I had to download over 350,000 XML files.&#13;
Passing all these as arguments will lead to an “Argument list too long” error from the command line itself.&#13;
My workaround is to accept the directory names as arguments:</p>&#13;
&#13;
<pre data-type="programlisting">parser.add_argument('-d',&#13;
                    '--dir',&#13;
                    metavar='DIR',&#13;
                    type=str,&#13;
                    nargs='+',&#13;
                    help='Input directories of FASTA/Q file(s)')</pre>&#13;
&#13;
<p>I then use Python to recursively search the directories for files.&#13;
For this code, I added <code>from pathlib import Path</code> so that I could use the <code>Path.rglob()</code> function:</p>&#13;
&#13;
<pre data-type="programlisting">files = []&#13;
for dirname in args.dir:&#13;
    if os.path.isdir(dirname):&#13;
        files.extend(list(Path(dirname).rglob('*')))&#13;
&#13;
if not files:&#13;
    parser.error('Found no files')&#13;
&#13;
return Args(files=files,&#13;
            file_format=args.format,&#13;
            percent=args.percent,&#13;
            max_reads=args.max,&#13;
            seed=args.seed,&#13;
            outdir=args.outdir)</pre>&#13;
&#13;
<p>The program can continue as before because Python does not have a problem storing several hundred thousand items in a list.</p>&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Going Further" data-type="sect1"><div class="sect1" id="idm45963626788504">&#13;
<h1>Going Further</h1>&#13;
&#13;
<p>This program always produces FASTA output. Add an <code>--outfmt</code> output format option so that you can specify the output format. Consider detecting the input file format and writing the output format in the same way as you did in <a data-type="xref" href="ch16.html#ch16">Chapter 16</a>. Be sure to add the appropriate tests to verify that your program works.</p>&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Review" data-type="sect1"><div class="sect1" id="idm45963626785368">&#13;
<h1>Review</h1>&#13;
&#13;
<ul>&#13;
<li>&#13;
<p>The <code>&gt;</code> record marker in a FASTA file is also the redirect operator in <code>bash</code>, so care must be taken to quote this value on the command line.</p>&#13;
</li>&#13;
<li>&#13;
<p>Deterministic approaches always produce the same output for the given input. Nondeterministic approaches produce different outputs for the same inputs.</p>&#13;
</li>&#13;
<li>&#13;
<p>The <code>random</code> module has functions to select numbers from various distributions, such as the uniform and normal distributions.</p>&#13;
</li>&#13;
</ul>&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
</div></section></body></html>