<html><head></head><body><section data-pdf-bookmark="Chapter 15. Seqmagique: Creating and Formatting Reports" data-type="chapter" epub:type="chapter"><div class="chapter" id="ch15">&#13;
<h1><span class="label">Chapter 15. </span>Seqmagique: Creating <span class="keep-together">and Formatting Reports</span></h1>&#13;
&#13;
&#13;
<p>Often in bioinformatics projects, you’ll find yourself staring at a directory full of sequence files, probably in FASTA or FASTQ format.<a data-primary="Seqmagick tool" data-secondary="about" data-type="indexterm" id="idm45963628769128"/><a data-primary="sequence file validation" data-secondary="about" data-type="indexterm" id="idm45963628768184"/><a data-primary="output from program" data-secondary="sequence file checking" data-seealso="Seqmagick tool" data-type="indexterm" id="idm45963628767224"/><a data-primary="FASTA format" data-secondary="sequence file checking" data-tertiary="about" data-type="indexterm" id="idm45963628766008"/><a data-primary="Seqmagique program created" data-secondary="about Seqmagick tool" data-type="indexterm" id="idm45963628764792"/>&#13;
You’ll probably want to start by getting an idea of the distribution of sequences in the files, such as how many are in each file and the average, minimum, and maximum lengths of the sequences.&#13;
You need to know if any files are corrupted—maybe they didn’t transfer completely from your sequencing center—or if any samples have far fewer reads, perhaps indicating a bad sequencing run that needs to be redone.&#13;
In this chapter, I’ll introduce some techniques for checking your sequence files using hashes and the <a href="https://oreil.ly/VI9gr">Seqmagick tool</a>.<a data-primary="Seqmagick tool" data-secondary="installing" data-tertiary="URL" data-type="indexterm" id="idm45963628762440"/><a data-primary="FASTA format" data-secondary="sequence file checking" data-tertiary="Seqmagick tool URL" data-type="indexterm" id="idm45963628761224"/><a data-primary="sequence file validation" data-secondary="Seqmagick tool" data-tertiary="URL" data-type="indexterm" id="idm45963628760008"/>&#13;
Then I’ll write a small utility to mimic part of Seqmagick to illustrate how to create formatted text tables.&#13;
This program serves as a template for any program that needs to process all the records in a given set of files and produce a table of summary statistics.</p>&#13;
&#13;
<p>You will learn:</p>&#13;
&#13;
<ul>&#13;
<li>&#13;
<p>How to install the <code>seqmagick</code> tool</p>&#13;
</li>&#13;
<li>&#13;
<p>How to use MD5 hashes</p>&#13;
</li>&#13;
<li>&#13;
<p>How to use <code>choices</code> in <code>argparse</code> to constrain arguments</p>&#13;
</li>&#13;
<li>&#13;
<p>How to use the <code>numpy</code> module</p>&#13;
</li>&#13;
<li>&#13;
<p>How to mock a filehandle</p>&#13;
</li>&#13;
<li>&#13;
<p>How to use the <code>tabulate</code> and <code>rich</code> modules to format output tables</p>&#13;
</li>&#13;
</ul>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Using Seqmagick to Analyze Sequence Files" data-type="sect1"><div class="sect1" id="UsingSeqmagick">&#13;
<h1>Using Seqmagick to Analyze Sequence Files</h1>&#13;
&#13;
<p><code>seqmagick</code> is a useful command-line utility for handling sequence files.&#13;
This should have been installed along with the other Python modules if you followed the setup instructions in the Preface.<a data-primary="Seqmagick tool" data-secondary="installing" data-type="indexterm" id="idm45963628747144"/><a data-primary="FASTA format" data-secondary="sequence file checking" data-tertiary="Seqmagick installation" data-type="indexterm" id="idm45963628746200"/><a data-primary="sequence file validation" data-secondary="Seqmagick tool" data-tertiary="installing" data-type="indexterm" id="idm45963628744984"/><a data-primary="Seqmagique program created" data-secondary="about Seqmagick tool" data-tertiary="analyzing sequence files" data-type="indexterm" id="idm45963628743752"/>&#13;
If not, you can install it with <code>pip</code>:<a data-primary="pip package installer" data-secondary="seqmagick" data-type="indexterm" id="idm45963628741880"/></p>&#13;
&#13;
<pre data-type="programlisting">$ python3 -m pip install seqmagick</pre>&#13;
&#13;
<p>If you run <strong><code>seqmagick --help</code></strong>, you’ll see the tool offers many options.<a data-primary="sequence file validation" data-secondary="Seqmagick tool" data-tertiary="analyzing sequence files" data-type="indexterm" id="idm45963628739144"/><a data-primary="Seqmagick tool" data-secondary="analyzing sequence files" data-type="indexterm" id="idm45963628737896"/><a data-primary="FASTA format" data-secondary="sequence file checking" data-tertiary="analyzing sequence files" data-type="indexterm" id="idm45963628736936"/>&#13;
I only want to focus on the <code>info</code> subcommand.<a data-primary="Seqmagick tool" data-secondary="info command on FASTA files" data-type="indexterm" id="idm45963628735080"/><a data-primary="FASTA format" data-secondary="sequence file checking" data-tertiary="Seqmagick info command" data-type="indexterm" id="idm45963628734120"/><a data-primary="sequence file validation" data-secondary="Seqmagick tool" data-tertiary="info command on FASTA files" data-type="indexterm" id="idm45963628732904"/>&#13;
I can run this on the test input FASTA files in the <em>15_seqmagique</em> directory like so:</p>&#13;
&#13;
<pre data-type="programlisting">$ cd 15_seqmagique&#13;
$ seqmagick info tests/inputs/*.fa&#13;
name                  alignment    min_len   max_len   avg_len  num_seqs&#13;
tests/inputs/1.fa     FALSE             50        50     50.00         1&#13;
tests/inputs/2.fa     FALSE             49        79     64.00         5&#13;
tests/inputs/empty.fa FALSE              0         0      0.00         0</pre>&#13;
&#13;
<p>In this exercise, you will create a program called <code>seqmagique.py</code> (this should be pronounced with an exaggerated French accent) that will mimic this output.&#13;
The point of the program is to provide a basic overview of the sequences in a given set of files so you can spot, for instance, a truncated or corrupted file.</p>&#13;
&#13;
<p>Start by copying the solution to <code>seqmagique.py</code> and requesting the usage:</p>&#13;
&#13;
<pre data-type="programlisting">$ cp solution1.py seqmagique.py&#13;
$ ./seqmagique.py -h&#13;
usage: seqmagique.py [-h] [-t table] FILE [FILE ...]&#13;
&#13;
Mimic seqmagick&#13;
&#13;
positional arguments:&#13;
  FILE                  Input FASTA file(s) <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO1-1" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO1-1"><img alt="1" src="assets/1.png"/></a>&#13;
&#13;
optional arguments:&#13;
  -h, --help            show this help message and exit&#13;
  -t table, --tablefmt table <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO1-2" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO1-2"><img alt="2" src="assets/2.png"/></a>&#13;
                        Tabulate table style (default: plain)</pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO1-1" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO1-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>The program accepts one or more input files which should be in FASTA format.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO1-2" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO1-2"><img alt="2" src="assets/2.png"/></a></dt>&#13;
<dd><p>This option controls the format of the output table.</p></dd>&#13;
</dl>&#13;
&#13;
<p>Run this program on the same files and note that the output is almost identical, except that I have omitted the <code>alignment</code> column:</p>&#13;
&#13;
<pre data-type="programlisting">$ ./seqmagique.py tests/inputs/*.fa&#13;
name                     min_len    max_len    avg_len    num_seqs&#13;
tests/inputs/1.fa             50         50      50.00           1&#13;
tests/inputs/2.fa             49         79      64.00           5&#13;
tests/inputs/empty.fa          0          0       0.00           0</pre>&#13;
&#13;
<p>The <code>--tablefmt</code> option controls how the output table is formatted.&#13;
This is the first program you’ll write that constrains the value to a given list.&#13;
To see this in action, use a bogus value like <code>blargh</code>:</p>&#13;
&#13;
<pre data-type="programlisting">$ ./seqmagique.py -t blargh tests/inputs/1.fa&#13;
usage: seqmagique.py [-h] [-t table] FILE [FILE ...]&#13;
seqmagique.py: error: argument -t/--tablefmt: invalid choice: 'blargh'&#13;
(choose from 'plain', 'simple', 'grid', 'pipe', 'orgtbl', 'rst',&#13;
 'mediawiki', 'latex', 'latex_raw', 'latex_booktabs')</pre>&#13;
&#13;
<p>Then try a different table format, such as <code>simple</code>:</p>&#13;
&#13;
<pre data-type="programlisting">$ ./seqmagique.py -t simple tests/inputs/*.fa&#13;
name                     min_len    max_len    avg_len    num_seqs&#13;
---------------------  ---------  ---------  ---------  ----------&#13;
tests/inputs/1.fa             50         50      50.00           1&#13;
tests/inputs/2.fa             49         79      64.00           5&#13;
tests/inputs/empty.fa          0          0       0.00           0</pre>&#13;
&#13;
<p>Run the program with other table styles and then try the test suite.&#13;
Next, I’ll talk about getting data for our program to analyze.</p>&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Checking Files Using MD5 Hashes" data-type="sect1"><div class="sect1" id="idm45963628709304">&#13;
<h1>Checking Files Using MD5 Hashes</h1>&#13;
&#13;
<p>The first step in most genomics projects will be transferring sequence files to some location where you can analyze them, and the first line of defense against data corruption is ensuring that the files were copied completely.<a data-primary="data corruption" data-see="sequence file validation" data-type="indexterm" id="idm45963628707784"/><a data-primary="Seqmagique program created" data-secondary="about simple file checks" data-type="indexterm" id="idm45963628706824"/>&#13;
The source of the files may be a sequencing center or a public repository like<a data-primary="data sources" data-secondary="GenBank URL" data-type="indexterm" id="idm45963628705624"/><a data-primary="data sources" data-secondary="Sequence Read Archive URL" data-type="indexterm" id="idm45963628704680"/><a data-primary="GenBank URL" data-type="indexterm" id="idm45963628703720"/><a data-primary="Sequence Read Archive (SRA) URL" data-type="indexterm" id="idm45963628703048"/><a data-primary="resources" data-secondary="public data repository URLs" data-type="indexterm" id="idm45963628702360"/> <a href="https://oreil.ly/2eaMj">GenBank</a> or the <a href="https://oreil.ly/kGNCv">Sequence Read Archive (SRA)</a>.&#13;
The files may arrive on a thumb drive, or you may download them from the internet.&#13;
If the latter, you may find that your connection drops, causing some files to be truncated or corrupted.&#13;
How can you find these types of errors?</p>&#13;
&#13;
<p>One way to check that your files are complete is to compare the file sizes locally with those on the server.<a data-primary="command line (Unix)" data-secondary="file size comparison" data-type="indexterm" id="idm45963628699032"/><a data-primary="Unix command line" data-secondary="file size comparison" data-type="indexterm" id="idm45963628698088"/><a data-primary="FASTA format" data-secondary="sequence file checking" data-tertiary="file size comparison" data-type="indexterm" id="idm45963628697144"/><a data-primary="sequence file validation" data-secondary="file size comparison" data-type="indexterm" id="idm45963628695928"/>&#13;
For instance, you can use the <strong><code>ls -l</code></strong> command to view the <em>long</em> listing of files where the file size, in bytes, is shown.&#13;
For large sequence files, this is going to be a very large number, and you will have to manually compare the file sizes from the source to the destination, which is tedious and prone to error.</p>&#13;
&#13;
<p>Another technique involves using a <em>hash</em> or <em>message digest</em> of the file, which is a signature of the file’s contents generated by a one-way cryptographic algorithm that creates a unique output for every possible input.<a data-primary="FASTA format" data-secondary="sequence file checking" data-tertiary="MD5 hashes to check files" data-type="indexterm" id="ch15-seq2"/><a data-primary="sequence file validation" data-secondary="MD5 hashes to check files" data-type="indexterm" id="ch15-seq3"/><a data-primary="MD5 hashes to check files" data-secondary="about" data-type="indexterm" id="ch15-seq"/><a data-primary="hash for file validation" data-see="MD5 hashes" data-type="indexterm" id="idm45963628688040"/><a data-primary="Seqmagique program created" data-secondary="about md5 hashes" data-type="indexterm" id="ch15-seq4"/>&#13;
Although there are many tools you can use to create a hash, I’ll focus on tools that use the MD5 algorithm.&#13;
This algorithm was originally developed in the context of cryptography and security, but researchers have since identified numerous flaws that now make it suitable only for purposes such as verifying data integrity.</p>&#13;
&#13;
<p>On macOS, I can use <strong><code>md5</code></strong> to generate a 128-bit hash value from the contents of the first test input file, like so:<a data-primary="Unix command line" data-secondary="md5 command for hash" data-type="indexterm" id="idm45963628684280"/><a data-primary="command line (Unix)" data-secondary="md5 command for hash" data-type="indexterm" id="idm45963628683336"/></p>&#13;
&#13;
<pre data-type="programlisting">$ md5 -r tests/inputs/1.fa&#13;
c383c386a44d83c37ae287f0aa5ae11d tests/inputs/1.fa</pre>&#13;
&#13;
<p>I can also use <strong><code>openssl</code></strong>:<a data-primary="Unix command line" data-secondary="openssl md5 for hash" data-type="indexterm" id="idm45963628680584"/><a data-primary="command line (Unix)" data-secondary="openssl md5 for hash" data-type="indexterm" id="idm45963628679640"/></p>&#13;
&#13;
<pre data-type="programlisting">$ openssl md5 tests/inputs/1.fa&#13;
MD5(tests/inputs/1.fa)= c383c386a44d83c37ae287f0aa5ae11d</pre>&#13;
&#13;
<p>On Linux, I use <strong><code>md5sum</code></strong>:<a data-primary="Linux virtual machine with Docker" data-secondary="md5sum for hash" data-type="indexterm" id="idm45963628676872"/></p>&#13;
&#13;
<pre data-type="programlisting">$ md5sum tests/inputs/1.fa&#13;
c383c386a44d83c37ae287f0aa5ae11d  tests/inputs/1.fa</pre>&#13;
&#13;
<p>As you can see, no matter the tool or platform, the hash value is the same for the same input file.&#13;
If I change even one bit of the input file, a different hash value will be generated.&#13;
Conversely, if I find another file that generates the same hash value, then the contents of the two files are identical.&#13;
For instance, the <em>empty.fa</em> file is a zero-length file I created for testing, and it has the following hash value:</p>&#13;
&#13;
<pre data-type="programlisting">$ md5 -r tests/inputs/empty.fa&#13;
d41d8cd98f00b204e9800998ecf8427e tests/inputs/empty.fa</pre>&#13;
&#13;
<p>If I use the <strong><code>touch foo</code></strong> command to create another empty file, I’ll find it has the same signature:</p>&#13;
&#13;
<pre data-type="programlisting">$ touch foo&#13;
$ md5 -r foo&#13;
d41d8cd98f00b204e9800998ecf8427e foo</pre>&#13;
&#13;
<p>It’s common for data providers to create a file of the checksums so that you can verify that your copies of the data are complete.&#13;
I created a <em>tests/inputs/checksums.md5</em> <span class="keep-together">like so:</span></p>&#13;
&#13;
<pre data-type="programlisting">$ cd tests/inputs&#13;
$ md5 -r *.fa &gt; checksums.md5</pre>&#13;
&#13;
<p>It has the following contents:</p>&#13;
&#13;
<pre data-type="programlisting">$ cat checksums.md5&#13;
c383c386a44d83c37ae287f0aa5ae11d 1.fa&#13;
863ebc53e28fdfe6689278e40992db9d 2.fa&#13;
d41d8cd98f00b204e9800998ecf8427e empty.fa</pre>&#13;
&#13;
<p>The <code>md5sum</code> tool has a <code>--check</code> option that I can use to automatically verify that the files match the checksums found in a given file.&#13;
The macOS <code>md5</code> tool does not have an option for this, but you can use <strong><code>brew install md5sha1sum</code></strong> to install an equivalent <code>md5sum</code> tool that can do this:</p>&#13;
&#13;
<pre data-type="programlisting">$ md5sum --check checksums.md5&#13;
1.fa: OK&#13;
2.fa: OK&#13;
empty.fa: OK</pre>&#13;
&#13;
<p>MD5 checksums present more complete and easier ways to verify data integrity than manually checking file sizes.<a data-startref="ch15-seq" data-type="indexterm" id="idm45963628663272"/><a data-startref="ch15-seq2" data-type="indexterm" id="idm45963628662600"/><a data-startref="ch15-seq3" data-type="indexterm" id="idm45963628661928"/><a data-startref="ch15-seq4" data-type="indexterm" id="idm45963628661256"/>&#13;
Although file digests are not directly part of this exercise, I feel it’s important to understand how to verify that you have complete and uncorrupted data before beginning any analyses.</p>&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Getting Started" data-type="sect1"><div class="sect1" id="idm45963628659928">&#13;
<h1>Getting Started</h1>&#13;
&#13;
<p>You should work in the <em>15_seqmagique</em> directory for this exercise.<a data-primary="MD5 hashes to check files" data-secondary="getting started" data-type="indexterm" id="idm45963628658168"/><a data-primary="Seqmagique program created" data-secondary="getting started" data-type="indexterm" id="idm45963628657208"/>&#13;
I’ll start the program as usual:</p>&#13;
&#13;
<pre data-type="programlisting">$ new.py -fp 'Mimic seqmagick' seqmagique.py&#13;
Done, see new script "seqmagique.py".</pre>&#13;
&#13;
<p>First I need to make the program accept one or more text files as positional parameters.&#13;
I also want to create an option to control the output table format.&#13;
Here is the code for that:</p>&#13;
&#13;
<pre data-type="programlisting">import argparse&#13;
from typing import NamedTuple, TextIO, List&#13;
&#13;
&#13;
class Args(NamedTuple):&#13;
    """ Command-line arguments """&#13;
    files: List[TextIO]&#13;
    tablefmt: str&#13;
&#13;
&#13;
def get_args() -&gt; Args:&#13;
    """Get command-line arguments"""&#13;
&#13;
    parser = argparse.ArgumentParser(&#13;
        description='Argparse Python script',&#13;
        formatter_class=argparse.ArgumentDefaultsHelpFormatter)&#13;
&#13;
    parser.add_argument('file', <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO2-1" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO2-1"><img alt="1" src="assets/1.png"/></a>&#13;
                        metavar='FILE',&#13;
                        type=argparse.FileType('rt'),&#13;
                        nargs='+',&#13;
                        help='Input FASTA file(s)')&#13;
&#13;
    parser.add_argument('-t',&#13;
                        '--tablefmt',&#13;
                        metavar='table',&#13;
                        type=str,&#13;
                        choices=[ <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO2-2" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO2-2"><img alt="2" src="assets/2.png"/></a>&#13;
                            'plain', 'simple', 'grid', 'pipe', 'orgtbl', 'rst',&#13;
                            'mediawiki', 'latex', 'latex_raw', 'latex_booktabs'&#13;
                        ],&#13;
                        default='plain',&#13;
                        help='Tabulate table style')&#13;
&#13;
    args = parser.parse_args()&#13;
&#13;
    return Args(args.file, args.tablefmt)</pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO2-1" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO2-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>Define a positional parameter for one or more readable text files.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO2-2" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO2-2"><img alt="2" src="assets/2.png"/></a></dt>&#13;
<dd><p>Define an option that uses <code>choices</code> to constrain the argument to a value in the list, making sure to define a reasonable <code>default</code> value.</p></dd>&#13;
</dl>&#13;
&#13;
<p>Using <code>choices</code> for the <code>--tablefmt</code> really saves you quite a bit of work in validating user input.&#13;
As shown in <a data-type="xref" href="#UsingSeqmagick">“Using Seqmagick to Analyze Sequence Files”</a>, a bad value for the table format option will trigger a useful error message.</p>&#13;
&#13;
<p>Modify the <code>main()</code> function to print the input filenames:</p>&#13;
&#13;
<pre data-type="programlisting">def main() -&gt; None:&#13;
    args = get_args()&#13;
    for fh in args.files:&#13;
        print(fh.name)</pre>&#13;
&#13;
<p>And verify that this works:</p>&#13;
&#13;
<pre data-type="programlisting">$ ./seqmagique.py tests/inputs/*.fa&#13;
tests/inputs/1.fa&#13;
tests/inputs/2.fa&#13;
tests/inputs/empty.fa</pre>&#13;
&#13;
<p>The goal is to iterate through each file and print the following:</p>&#13;
<dl>&#13;
<dt><code>name</code></dt>&#13;
<dd>&#13;
<p>The filename</p>&#13;
</dd>&#13;
<dt><code>min_len</code></dt>&#13;
<dd>&#13;
<p>The length of the shortest sequence</p>&#13;
</dd>&#13;
<dt><code>max_len</code></dt>&#13;
<dd>&#13;
<p>The length of the longest sequence</p>&#13;
</dd>&#13;
<dt><code>avg_len</code></dt>&#13;
<dd>&#13;
<p>The average/mean length of all the sequences</p>&#13;
</dd>&#13;
<dt><code>num_seqs</code></dt>&#13;
<dd>&#13;
<p>The number of sequences</p>&#13;
</dd>&#13;
</dl>&#13;
&#13;
<p>If you would like to have some real input files for the program, you can use the <a href="https://oreil.ly/Vmb0w"><code>fastq-dump</code> tool</a> from NCBI to download sequences from the study <a href="https://oreil.ly/aAGUA">“Planktonic Microbial Communities from North Pacific Subtropical Gyre”</a>:</p>&#13;
&#13;
<pre data-type="programlisting">$ fastq-dump --split-3 SAMN00000013 <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO3-1" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO3-1"><img alt="1" src="assets/1.png"/></a></pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO3-1" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO3-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>The <code>--split-3</code> option will ensure that paired-end reads are correctly split into forward/reverse/unpaired reads. The <code>SAMN00000013</code> string is the accession of <a href="https://oreil.ly/kBCQU">one of the samples</a> from the experiment.</p></dd>&#13;
</dl>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Formatting Text Tables Using tabulate()" data-type="sect2"><div class="sect2" id="idm45963628618008">&#13;
<h2>Formatting Text Tables Using tabulate()</h2>&#13;
&#13;
<p>The output of the program will be a text table formatted using the <code>tabulate()</code> function from that module.<a data-primary="Seqmagique program created" data-secondary="tabulate() to format tables" data-type="indexterm" id="idm45963628616184"/><a data-primary="MD5 hashes to check files" data-secondary="tabulate() to format tables" data-type="indexterm" id="idm45963628615208"/><a data-primary="output from program" data-secondary="tabulate() to format tables" data-type="indexterm" id="idm45963628614232"/>&#13;
Be sure to read the documentation:</p>&#13;
&#13;
<pre data-type="programlisting">&gt;&gt;&gt; from tabulate import tabulate&#13;
&gt;&gt;&gt; help(tabulate)</pre>&#13;
&#13;
<p>I need to define the headers for the table, and I decided to use the same ones as Seqmagick (minus the <code>alignment</code> column):</p>&#13;
&#13;
<pre data-type="programlisting">&gt;&gt;&gt; hdr = ['name', 'min_len', 'max_len', 'avg_len', 'num_seqs']</pre>&#13;
&#13;
<p>The first test file, <em>tests/inputs/1.fa</em>, has just one sequence of 50 bases, so the columns for this are as follows:</p>&#13;
&#13;
<pre data-type="programlisting">&gt;&gt;&gt; f1 = ['tests/inputs/1.fa', 50, 50, 50.00, 1]</pre>&#13;
&#13;
<p>The second test file, <em>tests/inputs/2.fa</em>, has five sequences ranging from 49 bases to 79 with an average length of 64 bases:</p>&#13;
&#13;
<pre data-type="programlisting">&gt;&gt;&gt; f2 = ['tests/inputs/2.fa', 49, 79, 64.00, 5]</pre>&#13;
&#13;
<p>The <code>tabulate()</code> function expects the table data to be passed positionally as a list of lists, and I can specify the <code>headers</code> as a keyword argument:</p>&#13;
&#13;
<pre data-type="programlisting">&gt;&gt;&gt; print(tabulate([f1, f2], headers=hdr))&#13;
name                 min_len    max_len    avg_len    num_seqs&#13;
-----------------  ---------  ---------  ---------  ----------&#13;
tests/inputs/1.fa         50         50         50           1&#13;
tests/inputs/2.fa         49         79         64           5</pre>&#13;
&#13;
<p>Alternatively, I can place the headers as the first row of data and indicate this is the location of the headers:</p>&#13;
&#13;
<pre data-type="programlisting">&gt;&gt;&gt; print(tabulate([hdr, f1, f2], headers='firstrow'))&#13;
name                 min_len    max_len    avg_len    num_seqs&#13;
-----------------  ---------  ---------  ---------  ----------&#13;
tests/inputs/1.fa         50         50         50           1&#13;
tests/inputs/2.fa         49         79         64           5</pre>&#13;
&#13;
<p>Note that the default table style for the <code>tabulate()</code> function is <code>simple</code>, but the <code>plain</code> format is what I need to match Seqmagick’s output.&#13;
I can set this with the <code>tablefmt</code> option:</p>&#13;
&#13;
<pre data-type="programlisting">&gt;&gt;&gt; print(tabulate([f1, f2], headers=hdr, tablefmt='plain'))&#13;
name                 min_len    max_len    avg_len    num_seqs&#13;
tests/inputs/1.fa         50         50         50           1&#13;
tests/inputs/2.fa         49         79         64           5</pre>&#13;
&#13;
<p>One other thing to note is that the values in the <code>avg_len</code> column are being shown as integers but should be formatted as floating-point numbers to two decimal places.&#13;
The <code>floatfmt</code> option controls this, using syntax similar to the f-string number formatting I’ve shown before:</p>&#13;
&#13;
<pre data-type="programlisting">&gt;&gt;&gt; print(tabulate([f1, f2], headers=hdr, tablefmt='plain', floatfmt='.2f'))&#13;
name                 min_len    max_len    avg_len    num_seqs&#13;
tests/inputs/1.fa         50         50      50.00           1&#13;
tests/inputs/2.fa         49         79      64.00           5</pre>&#13;
&#13;
<p>Your job is to process all the sequences in each file to find the statistics and print the final table.&#13;
This should be enough for you to solve the problem.&#13;
Don’t read ahead until you can pass all the tests.</p>&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Solutions" data-type="sect1"><div class="sect1" id="idm45963628596264">&#13;
<h1>Solutions</h1>&#13;
&#13;
<p>I’ll present two solutions that both show the file statistics but differ in the formatting of the output.&#13;
The first solution uses the <code>tabulate()</code> function to create an ASCII text table  and the second uses the <code>rich</code> module to create a fancier table sure to impress your labmates and principal investigator (PI).</p>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Solution 1: Formatting with tabulate()" data-type="sect2"><div class="sect2" id="idm45963628593656">&#13;
<h2>Solution 1: Formatting with tabulate()</h2>&#13;
&#13;
<p>For my solution, I first decided to write a <code>process()</code> function that would handle each input file.<a data-primary="Seqmagique program created" data-secondary="solution 1 tabulate() to format" data-type="indexterm" id="ch15-tab"/><a data-primary="MD5 hashes to check files" data-secondary="solution 1 tabulate() to format" data-type="indexterm" id="ch15-tab2"/><a data-primary="output from program" data-secondary="tabulate() to format tables" data-type="indexterm" id="ch15-tab3"/>&#13;
Whenever I approach a problem that needs to handle some list of items, I prefer to focus on how to handle just one of the items.&#13;
That is, rather than trying to find all the statistics for all the files, I first want to figure out how to find this information for just one file.</p>&#13;
&#13;
<p>My function needs to return the filename and the four metrics: minimum/maximum/average sequence lengths, plus the number of sequences.&#13;
Just as with the <code>Args</code> class, I like to create a type based on a <code>NamedTuple</code> for this so that I have a statically typed data structure that <code>mypy</code> can validate:</p>&#13;
&#13;
<pre data-type="programlisting">class FastaInfo(NamedTuple):&#13;
    """ FASTA file information """&#13;
    filename: str&#13;
    min_len: int&#13;
    max_len: int&#13;
    avg_len: float&#13;
    num_seqs: int</pre>&#13;
&#13;
<p>Now I can define a function that returns this data structure.&#13;
Note that I’m using the <code>numpy.mean()</code> function to get the average length.<a data-primary="NumPy module" data-secondary="mean() function" data-type="indexterm" id="idm45963628583816"/>&#13;
The <code>numpy</code> module offers many powerful mathematical operations to handle numeric<a data-primary="NumPy module" data-secondary="about" data-type="indexterm" id="idm45963628582216"/> data and is especially useful for multidimensional arrays and linear algebra functions.&#13;
When importing the dependencies, it’s common to import the <code>numpy</code> module with the alias <code>np</code>:<a data-primary="alias in Python" data-secondary="NumPy imported as np" data-type="indexterm" id="idm45963628580088"/><a data-primary="Python" data-secondary="aliasing" data-type="indexterm" id="idm45963628579144"/><a data-primary="NumPy module" data-secondary="importing" data-type="indexterm" id="idm45963628578200"/></p>&#13;
&#13;
<pre data-type="programlisting">import numpy as np&#13;
from tabulate import tabulate&#13;
from Bio import SeqIO</pre>&#13;
&#13;
<p>You can run <strong><code>help(np)</code></strong> in the REPL to read the documentation.&#13;
Here’s how I wrote this function:</p>&#13;
&#13;
<pre data-type="programlisting">def process(fh: TextIO) -&gt; FastaInfo: <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-1" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-1"><img alt="1" src="assets/1.png"/></a>&#13;
    """ Process a file """&#13;
&#13;
    if lengths := [len(rec.seq) for rec in SeqIO.parse(fh, 'fasta')]: <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-2" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-2"><img alt="2" src="assets/2.png"/></a>&#13;
        return FastaInfo(filename=fh.name, <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-3" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-3"><img alt="3" src="assets/3.png"/></a>&#13;
                         min_len=min(lengths), <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-4" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-4"><img alt="4" src="assets/4.png"/></a>&#13;
                         max_len=max(lengths), <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-5" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-5"><img alt="5" src="assets/5.png"/></a>&#13;
                         avg_len=round(float(np.mean(lengths)), 2), <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-6" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-6"><img alt="6" src="assets/6.png"/></a>&#13;
                         num_seqs=len(lengths)) <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-7" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-7"><img alt="7" src="assets/7.png"/></a>&#13;
&#13;
    return FastaInfo(filename=fh.name, <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-8" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-8"><img alt="8" src="assets/8.png"/></a>&#13;
                     min_len=0,&#13;
                     max_len=0,&#13;
                     avg_len=0,&#13;
                     num_seqs=0)</pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-1" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>The function accepts a filehandle and returns a <code>FastaInfo</code> object.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-2" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-2"><img alt="2" src="assets/2.png"/></a></dt>&#13;
<dd><p>Use a list comprehension to read all the sequences from the filehandle. Use the <code>len()</code> function to return the length of each sequence.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-3" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-3"><img alt="3" src="assets/3.png"/></a></dt>&#13;
<dd><p>The name of the file is available through the <code>fh.name</code> attribute.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-4" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-4"><img alt="4" src="assets/4.png"/></a></dt>&#13;
<dd><p>The <code>min()</code> function will return the minimum value.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-5" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-5"><img alt="5" src="assets/5.png"/></a></dt>&#13;
<dd><p>The <code>max()</code> function will return the maximum value.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-6" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-6"><img alt="6" src="assets/6.png"/></a></dt>&#13;
<dd><p>The <code>np.mean()</code> function will return the mean from a list of values. The <code>round()</code> function is used to round this floating-point value to two significant digits.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-7" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-7"><img alt="7" src="assets/7.png"/></a></dt>&#13;
<dd><p>The number of sequences is the length of the list.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-8" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO4-8"><img alt="8" src="assets/8.png"/></a></dt>&#13;
<dd><p>If there are no sequences, return zeros for all the values.</p></dd>&#13;
</dl>&#13;
&#13;
<p>As always, I want to write a unit test for this.<a data-primary="Seqmagique program created" data-secondary="solution 1 tabulate() to format" data-tertiary="unit test with mock filehandle" data-type="indexterm" id="idm45963628532264"/><a data-primary="MD5 hashes to check files" data-secondary="solution 1 tabulate() to format" data-tertiary="unit test with mock filehandle" data-type="indexterm" id="idm45963628531000"/>&#13;
While it’s true that the integration tests I wrote cover this part of the program, I want to show how you can write a unit test for a function that reads a file.<a data-primary="unit testing" data-secondary="function that reads a file" data-type="indexterm" id="idm45963628529432"/>&#13;
Rather than relying on actual files, I’ll create a <em>mock</em> or fake filehandle.<a data-primary="filehandle" data-secondary="mock filehandle" data-type="indexterm" id="idm45963628527832"/></p>&#13;
&#13;
<p>The first test file looks like this:</p>&#13;
&#13;
<pre data-type="programlisting">$ cat tests/inputs/1.fa&#13;
&gt;SEQ0&#13;
GGATAAAGCGAGAGGCTGGATCATGCACCAACTGCGTGCAACGAAGGAAT</pre>&#13;
&#13;
<p>I can use the <code>io.StringIO()</code> function to create an object that behaves like a <span class="keep-together">filehandle:</span><a data-primary="io.StringIO() for mock filehandle" data-type="indexterm" id="idm45963628524152"/><a data-primary="StringIO() for mock filehandle" data-type="indexterm" id="idm45963628523464"/></p>&#13;
&#13;
<pre data-type="programlisting">&gt;&gt;&gt; import io&#13;
&gt;&gt;&gt; f1 = '&gt;SEQ0\nGGATAAAGCGAGAGGCTGGATCATGCACCAACTGCGTGCAACGAAGGAAT\n' <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO5-1" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO5-1"><img alt="1" src="assets/1.png"/></a>&#13;
&gt;&gt;&gt; fh = io.StringIO(f1) <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO5-2" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO5-2"><img alt="2" src="assets/2.png"/></a>&#13;
&gt;&gt;&gt; for line in fh: <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO5-3" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO5-3"><img alt="3" src="assets/3.png"/></a>&#13;
...     print(line, end='') <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO5-4" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO5-4"><img alt="4" src="assets/4.png"/></a>&#13;
...&#13;
&gt;SEQ0&#13;
GGATAAAGCGAGAGGCTGGATCATGCACCAACTGCGTGCAACGAAGGAAT</pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO5-1" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO5-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>This is the data from the first input file.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO5-2" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO5-2"><img alt="2" src="assets/2.png"/></a></dt>&#13;
<dd><p>Create a mock filehandle.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO5-3" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO5-3"><img alt="3" src="assets/3.png"/></a></dt>&#13;
<dd><p>Iterate through the lines of the mock filehandle.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO5-4" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO5-4"><img alt="4" src="assets/4.png"/></a></dt>&#13;
<dd><p>Print the line which has a newline (<code>\n</code>), so use <code>end=''</code> to leave off an additional newline.</p></dd>&#13;
</dl>&#13;
&#13;
<p>There’s a slight problem, though, because the <code>process()</code> function calls the <code>fh.name</code> attribute to get the input filename, which will raise an exception:</p>&#13;
&#13;
<pre data-type="programlisting">&gt;&gt;&gt; fh.name&#13;
Traceback (most recent call last):&#13;
  File "&lt;stdin&gt;", line 1, in &lt;module&gt;&#13;
AttributeError: '_io.StringIO' object has no attribute 'name'</pre>&#13;
&#13;
<p>Luckily, there’s another way to create a mock filehandle using Python’s standard <span class="keep-together"><code>unittest</code></span> module.<a data-primary="unittest module for mock filehandle" data-type="indexterm" id="idm45963628497928"/>&#13;
While I favor the <code>pytest</code> module for almost everything I write, <span class="keep-together">the <code>unittest</code></span> module has been around for a long time and is another capable framework for writing and running tests.<a data-primary="pytest tool" data-secondary="about" data-type="indexterm" id="idm45963628495528"/>&#13;
In this case, I need to import the <a href="https://oreil.ly/EGvXh"><code>uni⁠t​test.mock.mock_open()</code> function.</a>&#13;
Here is how I can create a mock filehandle with the data from the first test file.&#13;
I use <code>read_data</code> to define the data that will be returned by the <code>fh.read()</code> method:</p>&#13;
&#13;
<pre data-type="programlisting">&gt;&gt;&gt; from unittest.mock import mock_open&#13;
&gt;&gt;&gt; fh = mock_open(read_data=f1)()&#13;
&gt;&gt;&gt; fh.read()&#13;
'&gt;SEQ0\nGGATAAAGCGAGAGGCTGGATCATGCACCAACTGCGTGCAACGAAGGAAT\n'</pre>&#13;
&#13;
<p>In the context of testing, I don’t care about the filename, only that this returns a string and does not throw an exception:</p>&#13;
&#13;
<pre data-type="programlisting">&gt;&gt;&gt; fh.name&#13;
&lt;MagicMock name='open().name' id='140349116126880'&gt;</pre>&#13;
&#13;
<p>While I often place my unit tests in the same modules as the functions they test, in this instance, I’d rather put this into a separate <code>unit.py</code> module to keep the main program shorter.&#13;
I wrote the test to handle an empty file, a file with one sequence, and a file with more than one sequence (which are also reflected in the three input test files).&#13;
Presumably, if the function works for these three cases, it should work for all others:</p>&#13;
&#13;
<pre data-type="programlisting">from unittest.mock import mock_open <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-1" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-1"><img alt="1" src="assets/1.png"/></a>&#13;
from seqmagique import process <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-2" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-2"><img alt="2" src="assets/2.png"/></a>&#13;
&#13;
&#13;
def test_process() -&gt; None:&#13;
    """ Test process """&#13;
&#13;
    empty = process(mock_open(read_data='')()) <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-3" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-3"><img alt="3" src="assets/3.png"/></a>&#13;
    assert empty.min_len == 0&#13;
    assert empty.max_len == 0&#13;
    assert empty.avg_len == 0&#13;
    assert empty.num_seqs == 0&#13;
&#13;
    one = process(mock_open(read_data='&gt;SEQ0\nAAA')()) <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-4" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-4"><img alt="4" src="assets/4.png"/></a>&#13;
    assert one.min_len == 3&#13;
    assert one.max_len == 3&#13;
    assert one.avg_len == 3&#13;
    assert one.num_seqs == 1&#13;
&#13;
    two = process(mock_open(read_data='&gt;SEQ0\nAAA\n&gt;SEQ1\nCCCC')()) <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-5" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-5"><img alt="5" src="assets/5.png"/></a>&#13;
    assert two.min_len == 3&#13;
    assert two.max_len == 4&#13;
    assert two.avg_len == 3.5&#13;
    assert two.num_seqs == 2</pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-1" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>Import the <code>mock_open()</code> function.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-2" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-2"><img alt="2" src="assets/2.png"/></a></dt>&#13;
<dd><p>Import the <code>process()</code> function I’m testing.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-3" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-3"><img alt="3" src="assets/3.png"/></a></dt>&#13;
<dd><p>A mock empty filehandle that should have zeros for all the values.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-4" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-4"><img alt="4" src="assets/4.png"/></a></dt>&#13;
<dd><p>A single sequence with three bases.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-5" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO6-5"><img alt="5" src="assets/5.png"/></a></dt>&#13;
<dd><p>A filehandle with two sequences of three and four bases.</p></dd>&#13;
</dl>&#13;
&#13;
<p class="pagebreak-before">Use <strong><code>pytest</code></strong> to run the tests:</p>&#13;
&#13;
<pre data-type="programlisting">$ pytest -xv unit.py&#13;
============================= test session starts ==============================&#13;
...&#13;
&#13;
unit.py::test_process PASSED                                             [100%]&#13;
&#13;
============================== 1 passed in 2.55s ===============================</pre>&#13;
<aside data-type="sidebar" epub:type="sidebar"><div class="sidebar" id="idm45963628460184">&#13;
<h5>Where to Place Unit Tests</h5>&#13;
<p>Note that the preceding <code>unit.py</code> module imports the <code>process()</code> function from the <code>seqmagique.py</code> module, so both modules need to be in the same directory.<a data-primary="unit testing" data-secondary="function that reads a file" data-tertiary="where to place unit tests" data-type="indexterm" id="idm45963628457464"/><a data-primary="Seqmagique program created" data-secondary="solution 1 tabulate() to format" data-tertiary="where to place unit tests" data-type="indexterm" id="idm45963628456216"/><a data-primary="MD5 hashes to check files" data-secondary="solution 1 tabulate() to format" data-tertiary="where to place unit tests" data-type="indexterm" id="idm45963628454952"/>&#13;
If I were to move <code>unit.py</code> to the <em>tests</em> directory, then <code>pytest</code> would break.&#13;
I encourage you to try the following and note the errors.</p>&#13;
&#13;
<p>You should get a notice like <em>ModuleNotFoundError: No module named seqmagique</em>:</p>&#13;
&#13;
<pre data-type="programlisting">$ cp unit.py tests&#13;
$ pytest -xv tests/unit.py</pre>&#13;
&#13;
<p>As noted <a href="https://oreil.ly/um7N7">in the documentation</a>, I must invoke <code>pytest</code> as follows to add the current directory to <code>sys.path</code> so that <code>seqmagique.py</code> can be found:</p>&#13;
&#13;
<pre data-type="programlisting">$ python3 -m pytest -xv tests/unit.py</pre>&#13;
&#13;
<p>Placing <code>unit.py</code> in the same directory as the code it’s testing is slightly more convenient as it means I can run the shorter <code>pytest</code> command, but grouping all the tests into the <em>tests</em> directory is tidier.&#13;
I would normally prefer to have this module as <em>tests/unit_test.py</em> so that <code>pytest</code> will automatically discover it, and I would use a <code>make</code> target to run the longer invocation.&#13;
Mostly I just want you to be aware of different ways to organize your code and tests.</p>&#13;
</div></aside>&#13;
&#13;
<p>Here is how I use my <code>process()</code> function in <code>main()</code>:<a data-primary="formatting tables" data-secondary="tabulate()" data-type="indexterm" id="idm45963628442200"/><a data-primary="table formatting" data-secondary="tabulate()" data-type="indexterm" id="idm45963628441256"/></p>&#13;
&#13;
<pre data-type="programlisting">def main() -&gt; None:&#13;
    args = get_args()&#13;
    data = [process(fh) for fh in args.files] <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO7-1" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO7-1"><img alt="1" src="assets/1.png"/></a>&#13;
    hdr = ['name', 'min_len', 'max_len', 'avg_len', 'num_seqs'] <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO7-2" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO7-2"><img alt="2" src="assets/2.png"/></a>&#13;
    print(tabulate(data, tablefmt=args.tablefmt, headers=hdr, floatfmt='.2f')) <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO7-3" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO7-3"><img alt="3" src="assets/3.png"/></a></pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO7-1" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO7-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>Process all the input files into a list of <code>FastaInfo</code> objects (tuples).</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO7-2" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO7-2"><img alt="2" src="assets/2.png"/></a></dt>&#13;
<dd><p>Define the table headers.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO7-3" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO7-3"><img alt="3" src="assets/3.png"/></a></dt>&#13;
<dd><p>Use the <code>tabulate()</code> function to print a formatted output table.</p></dd>&#13;
</dl>&#13;
&#13;
<p class="pagebreak-before">To test this program, I run it with the following inputs:</p>&#13;
&#13;
<ul>&#13;
<li>&#13;
<p>The empty file</p>&#13;
</li>&#13;
<li>&#13;
<p>The file with one sequence</p>&#13;
</li>&#13;
<li>&#13;
<p>The file with two sequences</p>&#13;
</li>&#13;
<li>&#13;
<p>All the input files</p>&#13;
</li>&#13;
</ul>&#13;
&#13;
<p>To start, I run all these with the default table style.&#13;
Then I need to verify that all 10 of the table styles are created correctly.&#13;
Combining all the possible test inputs with all the table styles creates a high degree of <em>cyclomatic complexity</em>—the number of different ways that parameters can be combined.</p>&#13;
&#13;
<p>To test this, I first need to manually verify that my program is working correctly.&#13;
Then I need to generate sample outputs for each of the combinations I intend to test.&#13;
I wrote the following <code>bash</code> script to create an <em>out</em> file for a given combination of an input file and possibly a table style:</p>&#13;
&#13;
<pre data-type="programlisting">$ cat mk-outs.sh&#13;
#!/usr/bin/env bash&#13;
&#13;
PRG="./seqmagique.py" <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-1" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-1"><img alt="1" src="assets/1.png"/></a>&#13;
DIR="./tests/inputs" <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-2" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-2"><img alt="2" src="assets/2.png"/></a>&#13;
INPUT1="${DIR}/1.fa" <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-3" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-3"><img alt="3" src="assets/3.png"/></a>&#13;
INPUT2="${DIR}/2.fa"&#13;
EMPTY="${DIR}/empty.fa"&#13;
&#13;
$PRG $INPUT1 &gt; "${INPUT1}.out" <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-4" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-4"><img alt="4" src="assets/4.png"/></a>&#13;
$PRG $INPUT2 &gt; "${INPUT2}.out"&#13;
$PRG $EMPTY &gt; "${EMPTY}.out"&#13;
$PRG $INPUT1 $INPUT2 $EMPTY &gt; "$DIR/all.fa.out"&#13;
&#13;
STYLES="plain simple grid pipe orgtbl rst mediawiki latex latex_raw&#13;
 latex_booktabs"&#13;
&#13;
for FILE in $INPUT1 $INPUT2; do <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-5" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-5"><img alt="5" src="assets/5.png"/></a>&#13;
    for STYLE in $STYLES; do&#13;
        $PRG -t $STYLE $FILE &gt; "$FILE.${STYLE}.out"&#13;
    done&#13;
done&#13;
&#13;
echo Done.</pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-1" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>The program being tested.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-2" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-2"><img alt="2" src="assets/2.png"/></a></dt>&#13;
<dd><p>The directory for the input files.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-3" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-3"><img alt="3" src="assets/3.png"/></a></dt>&#13;
<dd><p>The input files.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-4" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-4"><img alt="4" src="assets/4.png"/></a></dt>&#13;
<dd><p>Run the program using the three input files and the default table style.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-5" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO8-5"><img alt="5" src="assets/5.png"/></a></dt>&#13;
<dd><p>Run the program with two of the input files and all the table styles.</p></dd>&#13;
</dl>&#13;
&#13;
<p>The tests in <em>tests/seqmagique_test.py</em> will run the program with a given file and will compare the output to one of the <em>out</em> files in the <em>tests/inputs</em> directory.&#13;
At the top of this module, I define the input and output files like so:</p>&#13;
&#13;
<pre data-type="programlisting">TEST1 = ('./tests/inputs/1.fa', './tests/inputs/1.fa.out')</pre>&#13;
&#13;
<p>I define a <code>run()</code> function in the module to run the program with the input file and compare the actual output to the expected output.&#13;
This is a basic pattern you could copy for testing any program’s output:</p>&#13;
&#13;
<pre data-type="programlisting">def run(input_file: str, expected_file: str) -&gt; None:&#13;
    """ Runs on command-line input """&#13;
&#13;
    expected = open(expected_file).read().rstrip() <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO9-1" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO9-1"><img alt="1" src="assets/1.png"/></a>&#13;
    rv, out = getstatusoutput(f'{RUN} {input_file}') <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO9-2" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO9-2"><img alt="2" src="assets/2.png"/></a>&#13;
    assert rv == 0 <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO9-3" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO9-3"><img alt="3" src="assets/3.png"/></a>&#13;
    assert out == expected <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO9-4" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO9-4"><img alt="4" src="assets/4.png"/></a></pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO9-1" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO9-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>Read the expected output from the file.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO9-2" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO9-2"><img alt="2" src="assets/2.png"/></a></dt>&#13;
<dd><p>Run the program with the given input file using the default table style.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO9-3" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO9-3"><img alt="3" src="assets/3.png"/></a></dt>&#13;
<dd><p>Check that the return value is <code>0</code>.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO9-4" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO9-4"><img alt="4" src="assets/4.png"/></a></dt>&#13;
<dd><p>Check that the output was the expected value.</p></dd>&#13;
</dl>&#13;
&#13;
<p>I use it like so:</p>&#13;
&#13;
<pre data-type="programlisting">def test_input1() -&gt; None:&#13;
    """ Runs on command-line input """&#13;
&#13;
    run(*TEST1) <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO10-1" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO10-1"><img alt="1" src="assets/1.png"/></a></pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO10-1" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO10-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>Splat the tuple to pass the two values positionally to the <code>run()</code> function.</p></dd>&#13;
</dl>&#13;
&#13;
<p>The test suite also checks the table styles:</p>&#13;
&#13;
<pre data-type="programlisting">def test_styles() -&gt; None:&#13;
    """ Test table styles """&#13;
&#13;
    styles = [ <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-1" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-1"><img alt="1" src="assets/1.png"/></a>&#13;
        'plain', 'simple', 'grid', 'pipe', 'orgtbl', 'rst', 'mediawiki',&#13;
        'latex', 'latex_raw', 'latex_booktabs'&#13;
    ]&#13;
&#13;
    for file in [TEST1[0], TEST2[0]]: <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-2" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-2"><img alt="2" src="assets/2.png"/></a>&#13;
        for style in styles: <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-3" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-3"><img alt="3" src="assets/3.png"/></a>&#13;
            expected_file = file + '.' + style + '.out' <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-4" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-4"><img alt="4" src="assets/4.png"/></a>&#13;
            assert os.path.isfile(expected_file) <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-5" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-5"><img alt="5" src="assets/5.png"/></a>&#13;
            expected = open(expected_file).read().rstrip() <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-6" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-6"><img alt="6" src="assets/6.png"/></a>&#13;
            flag = '--tablefmt' if random.choice([0, 1]) else '-t' <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-7" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-7"><img alt="7" src="assets/7.png"/></a>&#13;
            rv, out = getstatusoutput(f'{RUN} {flag} {style} {file}') <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-8" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-8"><img alt="8" src="assets/8.png"/></a>&#13;
            assert rv == 0 <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-9" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-9"><img alt="9" src="assets/9.png"/></a>&#13;
            assert out == expected</pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-1" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>Define a list of all possible styles.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-2" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-2"><img alt="2" src="assets/2.png"/></a></dt>&#13;
<dd><p>Use the two nonempty files.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-3" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-3"><img alt="3" src="assets/3.png"/></a></dt>&#13;
<dd><p>Iterate through each style.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-4" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-4"><img alt="4" src="assets/4.png"/></a></dt>&#13;
<dd><p>The output file is the name of the input file plus the style and the extension <em>.out</em>.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-5" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-5"><img alt="5" src="assets/5.png"/></a></dt>&#13;
<dd><p>Check that the file exists.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-6" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-6"><img alt="6" src="assets/6.png"/></a></dt>&#13;
<dd><p>Read the expected value from the file.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-7" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-7"><img alt="7" src="assets/7.png"/></a></dt>&#13;
<dd><p>Randomly choose the short or long flag to test.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-8" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-8"><img alt="8" src="assets/8.png"/></a></dt>&#13;
<dd><p>Run the program with the flag option, style, and file.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-9" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO11-9"><img alt="9" src="assets/9.png"/></a></dt>&#13;
<dd><p>Ensure that the program ran without error and produces the correct output.</p></dd>&#13;
</dl>&#13;
&#13;
<p>If I make a change such that the program no longer creates the same output as before, these tests should catch it.&#13;
This is a <em>regression</em> test, where I am comparing how a program works now to how it previously worked.&#13;
That is, a failure to produce the same output would be considered a regression.&#13;
While my test suite isn’t completely exhaustive, it covers enough combinations that I feel confident the program is correct.<a data-startref="ch15-tab" data-type="indexterm" id="idm45963628313208"/><a data-startref="ch15-tab2" data-type="indexterm" id="idm45963628312536"/><a data-startref="ch15-tab3" data-type="indexterm" id="idm45963628311864"/></p>&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Solution 2: Formatting with rich" data-type="sect2"><div class="sect2" id="idm45963628310872">&#13;
<h2>Solution 2: Formatting with rich</h2>&#13;
&#13;
<p>In this second solution I want to show a different way to create the output table, using the <code>rich</code> module to track the processing of the input files and make a fancier output table.<a data-primary="Seqmagique program created" data-secondary="solution 2 rich to format" data-type="indexterm" id="idm45963628308968"/><a data-primary="MD5 hashes to check files" data-secondary="solution 2 rich to format" data-type="indexterm" id="idm45963628307992"/><a data-primary="output from program" data-secondary="rich to format tables" data-type="indexterm" id="idm45963628307016"/><a data-primary="rich to format tables" data-type="indexterm" id="idm45963628306072"/><a data-primary="formatting tables" data-secondary="rich module" data-type="indexterm" id="idm45963628305400"/><a data-primary="table formatting" data-secondary="rich module" data-type="indexterm" id="idm45963628304456"/>&#13;
<a data-type="xref" href="#fig_15.1">Figure 15-1</a> shows how the output looks.</p>&#13;
&#13;
<figure><div class="figure" id="fig_15.1">&#13;
<img alt="mpfb 1501" src="assets/mpfb_1501.png"/>&#13;
<h6><span class="label">Figure 15-1. </span>The progress indicator and output table using the <code>rich</code> module are fancier</h6>&#13;
</div></figure>&#13;
&#13;
<p>I still process the files in the same way, so the only difference is in creating the output.&#13;
I first need to import the needed functions:</p>&#13;
&#13;
<pre data-type="programlisting">from rich.console import Console&#13;
from rich.progress import track&#13;
from rich.table import Table, Column</pre>&#13;
&#13;
<p>Here is how I use these:</p>&#13;
&#13;
<pre data-type="programlisting">def main() -&gt; None:&#13;
    args = get_args()&#13;
&#13;
    table = Table('Name', <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-1" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-1"><img alt="1" src="assets/1.png"/></a>&#13;
                  Column(header='Min. Len', justify='right'),&#13;
                  Column(header='Max. Len', justify='right'),&#13;
                  Column(header='Avg. Len', justify='right'),&#13;
                  Column(header='Num. Seqs', justify='right'),&#13;
                  header_style="bold black")&#13;
&#13;
    for fh in track(args.file): <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-2" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-2"><img alt="2" src="assets/2.png"/></a>&#13;
        file = process(fh) <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-3" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-3"><img alt="3" src="assets/3.png"/></a>&#13;
        table.add_row(file.filename, str(file.min_len), str(file.max_len), <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-4" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-4"><img alt="4" src="assets/4.png"/></a>&#13;
                      str(file.avg_len), str(file.num_seqs))&#13;
&#13;
    console = Console() <a class="co" href="#callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-5" id="co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-5"><img alt="5" src="assets/5.png"/></a>&#13;
    console.print(table)</pre>&#13;
<dl class="calloutlist">&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-1" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-1"><img alt="1" src="assets/1.png"/></a></dt>&#13;
<dd><p>Create the table to hold the data. The Name column is a standard, left-justified string field. All the others need to be right-justified and require a custom <span class="keep-together"><code>Column</code></span> object.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-2" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-2"><img alt="2" src="assets/2.png"/></a></dt>&#13;
<dd><p>Iterate through each filehandle using the <code>track()</code> function to create a progress bar for the user.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-3" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-3"><img alt="3" src="assets/3.png"/></a></dt>&#13;
<dd><p>Process the file to get the statistics.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-4" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-4"><img alt="4" src="assets/4.png"/></a></dt>&#13;
<dd><p>Add the file’s statistics to the table. Note that all the values must be strings.</p></dd>&#13;
<dt><a class="co" href="#co_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-5" id="callout_seqmagique__creating__span_class__keep_together__and_formatting_reports__span__CO12-5"><img alt="5" src="assets/5.png"/></a></dt>&#13;
<dd><p>Create a <code>Console</code> object, and use it to print the output.</p></dd>&#13;
</dl>&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section data-pdf-bookmark="Going Further" data-type="sect1"><div class="sect1" id="idm45963628270264">&#13;
<h1>Going Further</h1>&#13;
&#13;
<p>The <code>seqmagick</code> tool has many other useful options.&#13;
Implement your own versions of as many as you can.</p>&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
<section class="pagebreak-before" data-pdf-bookmark="Review" data-type="sect1"><div class="sect1" id="idm45963628268136">&#13;
<h1>Review</h1>&#13;
&#13;
<p>Key points from this chapter:</p>&#13;
&#13;
<ul>&#13;
<li>&#13;
<p>The <code>seqmagick</code> tool provides many ways to examine sequence files.</p>&#13;
</li>&#13;
<li>&#13;
<p>There are many ways to verify that your input files are complete and not corrupted, from examining file sizes to using message digests such as MD5 hashes.</p>&#13;
</li>&#13;
<li>&#13;
<p>The <code>choices</code> option for <code>argparse</code> parameters will force the user to select a value from a given list.</p>&#13;
</li>&#13;
<li>&#13;
<p>The <code>tabulate</code> and <code>rich</code> modules can create text tables of data.</p>&#13;
</li>&#13;
<li>&#13;
<p>The <code>numpy</code> module is useful for many mathematical operations.</p>&#13;
</li>&#13;
<li>&#13;
<p>The <code>io.StringIO()</code> and <code>unittest.mock.mock_open()</code> functions offer two ways to mock a filehandle for testing.</p>&#13;
</li>&#13;
<li>&#13;
<p>Regression testing verifies that a program continues to work as it did before.</p>&#13;
</li>&#13;
</ul>&#13;
</div></section>&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
&#13;
</div></section></body></html>